http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/AwaitCloseChannelPoolMap.java

```diff
@@ -162,42 +166,43 @@ private Bootstrap createBootstrap(URI poolKey) {
     }
 
 
-    private boolean useProxyForHost(URI remoteAddr) {
+    private boolean shouldUseProxyForHost(URI remoteAddr) {
         if (proxyConfiguration == null) {
             return false;
         }
 
-        String remoteHost = StringUtils.lowerCase(remoteAddr.getHost());
-        for (String hostPattern : proxyConfiguration.nonProxyHosts()) {
-            if (remoteHost.matches(hostPattern)) {
-                return false;
-            }
-        }
 
-        return true;
+        return shouldProxyForHostCache.computeIfAbsent(remoteAddr, (uri) ->
+           proxyConfiguration.nonProxyHosts().stream().noneMatch(h -> uri.getHost().matches(h))
+        );
     }
 
     private String bootstrapHost(URI remoteHost) {
-        if (useProxyForHost(remoteHost)) {
+        if (shouldUseProxyForHost(remoteHost)) {
             return proxyConfiguration.host();
         }
         return remoteHost.getHost();
     }
 
     private int bootstrapPort(URI remoteHost) {
-        if (useProxyForHost(remoteHost)) {
+        if (shouldUseProxyForHost(remoteHost)) {
             return proxyConfiguration.port();
         }
         return remoteHost.getPort();
     }
 
     private URI proxyAddress(URI remoteHost) {
-        if (!useProxyForHost(remoteHost)) {
+        if (!shouldUseProxyForHost(remoteHost)) {
             return null;
         }
 
+        String scheme = proxyConfiguration.scheme();
+        if (scheme == null) {
+            scheme = "http";
+        }
+
         try {
-            return new URI(proxyConfiguration.scheme(), null, proxyConfiguration.host(), proxyConfiguration.port(), null, null,
+            return new URI(scheme, null, proxyConfiguration.host(), proxyConfiguration.port(), null, null,
```

Comment:
Style: I like to annotate nulls in constructors like this with comments. Totally optional suggestion.